crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -108,26 +149,67 @@ impl RecursorDnsHandle {
                             break 'max_forward;
                         };
 
-                        debug!("ns forwarded to {}", name);
+                        debug!(
+                            "ns for {} forwarded to {} via SOA record",
+                            query.name(),
+                            name
+                        );
                         zone = name.clone();
                     }
                     _ => return Err(e),
                 },
             }
         }
 
-        let ns = ns.ok_or_else(|| Error::from(format!("no nameserver found for {zone}")))?;
+        let mut ns = ns.ok_or_else(|| Error::from(format!("no nameserver found for {zone}")))?;
         debug!("found zone {} for {}", ns.zone(), query);
 
-        let response = self
+        match self
             .lookup(query.clone(), ns, request_time, query_has_dnssec_ok)
-            .await?;
-
-        // RFC 4035 section 3.2.1 if DO bit not set, strip DNSSEC records unless
-        // explicitly requested
-        let lookup = super::maybe_strip_dnssec_records(query_has_dnssec_ok, response, query);
-
-        Ok(lookup)
+            .await
+        {
+            Ok(response) => {
+                // RFC 4035 section 3.2.1 if DO bit not set, strip DNSSEC records unless
+                // explicitly requested
+                Ok(super::maybe_strip_dnssec_records(
+                    query_has_dnssec_ok,
+                    response,
+                    query,
+                ))
+            }
+            Err(e) => {
+                match e.kind() {
+                    // ErrorKind::ForwardNS is mapped from ProtoError(NoRecordsFound) when an authoritative
+                    // nameserver returns an empty answers sections (NoRecordsFound) and one or more
+                    // nameserver records in the nameservers section.  We build a new NS Pool based on those
+                    // records and call resolve against that NS pool.
+                    ErrorKind::ForwardNS(referral_ns) => {
```

Comment:
This feels like maybe it should be moved into a separate function, but I'm not going to be a stickler for that. Primarily I would do that if we add some unit tests for this section. I think we can merge as is until we have a good unit testing framework for the recursor.

Reply:
Sounds good. It looks like this change is causing three failures in the conformance test suite, all related to dnssec. One I've identified the cause of one (duplicate calls to base_name for DS records), but still need to look at the other two.